data_diff/dbt.py

```diff
@@ -0,0 +1,313 @@
+from dbt.config.renderer import ProfileRenderer
+
+CONST_RUN_RESULTS_PATH = "/target/run_results.json"
+CONST_MANIFEST_PATH = "/target/manifest.json"
+CONST_PROJECT_FILE = "/dbt_project.yml"
+CONST_PROFILES_FILE = "/profiles.yml"
+CONST_DBT_V_ABOVE = "1.0.0"
+CONST_DBT_V_BELOW = "1.4.0"
+
+
+@dataclass
+class DiffVars:
+    dev_path: List[str]
+    prod_path: List[str]
+    primary_keys: List[str]
+    datasource_id: str
+    connection: dict
+
+
+class DbtDiffer:
+    @classmethod
+    def diff(self, profiles_dir_override: str = None, project_dir_override: str = None, is_cloud: bool = False):
+        dbt_parser = DbtParser(profiles_dir_override, project_dir_override, is_cloud)
+        models = dbt_parser.get_models()
+        dbt_parser.set_project_dict()
+        datadiff_variables = dbt_parser.get_datadiff_variables()
+        config_prod_database = datadiff_variables.get("prod_database")
+        config_prod_schema = datadiff_variables.get("prod_schema")
+        datasource_id = datadiff_variables.get("datasource_id")
+
+        if not is_cloud:
+            dbt_parser.set_connection()
+
+        assert (
+            config_prod_database is not None or config_prod_schema is not None
+        ), "Expected a value for prod_database: or prod_schema: under \nvars:\n  data_diff: "
+
+        for model in models:
+            diff_vars = self._get_diff_vars(dbt_parser, config_prod_database, config_prod_schema, model, datasource_id)
+
+            if is_cloud and len(diff_vars.primary_keys) > 0:
+                self._cloud_diff(diff_vars)
+            elif is_cloud:
+                rich.print(
+                    f"[red]"
+                    + ".".join(diff_vars.dev_path)
+                    + " <> "
+                    + ".".join(diff_vars.prod_path)
+                    + "[/] \n"
+                    + "Skipped due to missing primary-key tag\n"
+                )
+
+            if not is_cloud and len(diff_vars.primary_keys) == 1:
+                self._local_diff(diff_vars)
+            elif not is_cloud:
+                rich.print(
+                    f"[red]"
+                    + ".".join(diff_vars.dev_path)
+                    + " <> "
+                    + ".".join(diff_vars.prod_path)
+                    + "[/] \n"
+                    + "Skipped due to missing primary-key tag or multi-column primary-key (unsupported for non --cloud diffs)\n"
+                )
+
+        rich.print("Diffs Complete!")
+
+    @staticmethod
+    def _get_diff_vars(dbt_parser, config_prod_database, config_prod_schema, model, datasource_id) -> DiffVars:
+        dev_database = model.database
+        dev_schema = model.schema_
+        primary_keys = dbt_parser.get_primary_keys(model)
+
+        if config_prod_database is None:
+            prod_database = dev_database
+        else:
+            prod_database = config_prod_database
+
+        if config_prod_schema is None:
+            prod_schema = dev_schema
```

Comment:
Just to clarify: Somewhere above in the PR, it says that the schema should not be specified if it is multi-schema diff — even for production database. However, here it tries to use the dev schema as the prod schema even if the prod database is specified properly but with not schema. Is this intended?

Reply:
Yes this is intended. What this does is make the schema dynamic for each model rather than fixed

If the dbt project uses database environments (so config_prod_schema is None), we are assuming that the same schema name being used in dev_db and prod_db are equivalent for each model

This assumption isn't always going to be correct, especially if they use [custom schemas](https://docs.getdbt.com/docs/build/custom-schemas) but this covers most cases in the absence of a prod manifest.json